src/crimson/os/seastore/journal/circular_bounded_journal.h

```diff
@@ -0,0 +1,286 @@
+    // TODO
+    return seastar::now();
+  }
+
+  replay_ret replay(delta_handler_t &&delta_handler);
+
+  open_for_write_ertr::future<> _open_device(const std::string path);
+
+  struct cbj_header_t;
+  using write_ertr = crimson::errorator<
+    crimson::ct_error::input_output_error,
+    crimson::ct_error::erange>;
+  /*
+   * append_record
+   *
+   * append data to current write position of CBJournal
+   *
+   * @param bufferlist to write
+   * @param rbm_abs_addr where data is written
+   *
+   */
+  write_ertr::future<> append_record(ceph::bufferlist bl, rbm_abs_addr addr);
+  /*
+   * device_write_bl
+   *
+   * @param device address to write
+   * @param bufferlist to write
+   *
+   */
+  write_ertr::future<> device_write_bl(rbm_abs_addr offset, ceph::bufferlist &bl);
+
+  using read_ertr = crimson::errorator<
+    crimson::ct_error::input_output_error,
+    crimson::ct_error::invarg,
+    crimson::ct_error::enoent,
+    crimson::ct_error::erange>;
+  using read_record_ertr = read_ertr;
+  using read_record_ret = read_record_ertr::future<
+	std::optional<std::pair<record_group_header_t, bufferlist>>
+	>;
+  using read_super_ertr = read_ertr;
+  using read_super_ret = read_super_ertr::future<
+	std::optional<std::pair<cbj_header_t, bufferlist>>
+	>;
+  /*
+   * read_record
+   *
+   * read record from given address
+   *
+   * @param paddr_t to read
+   *
+   */
+  read_record_ret read_record(paddr_t offset);
+  /*
+   * read_super
+   *
+   * read super block from given absolute address
+   *
+   * @param absolute address
+   *
+   */
+  read_super_ret read_super(rbm_abs_addr start);
+
+  ceph::bufferlist encode_super();
+
+  using mkfs_ertr = crimson::errorator<
+    crimson::ct_error::input_output_error,
+    crimson::ct_error::invarg
+  >;
+  using mkfs_ret = mkfs_ertr::future<>;
+  mkfs_ret mkfs(mkfs_config_t& config);
+
+
+  /**
+   * CBJournal structure
+   *
+   * +-------------------------------------------------------+
+   * |   header    | record | record | record | record | ... |
+   * +-------------------------------------------------------+
+   *               ^-----------block aligned-----------------^
+   * <----fixed---->
+   */
+
+
+  /**
+   *
+   * CBJournal write
+   *
+   * NVMe will support a large block write (< 512KB) with atomic write unit command.
+   * With this command, we expect that the most of incoming data can be stored
+   * as a single write call, which has lower overhead than existing
+   * way that uses a combination of system calls such as write() and sync().
+   *
+   */
+
+  struct cbj_header_t {
+    uint64_t magic;
+    uuid_d uuid;
+    uint64_t block_size; // aligned with block_size
+    uint64_t size;   // max length of journal
+    uint64_t used_size;  // current used_size of journal
+    uint32_t error;      // reserved
+
+    rbm_abs_addr start_offset;      // start offset of CBJournal
+    rbm_abs_addr last_committed_record_base;
+    rbm_abs_addr written_to;
+    rbm_abs_addr applied_to;
```

Comment:
I'm not sure what applied_to is meant to denote.

Reply:
This represents the cursor which indicates the last record---moved to RBM most recently---position in CircularBoundedJournal. The cleaner will take the role---move records from CircularBoundedJournal to RBM, but the cleaner for CBjournal is not implemented yet. I added this to confirm that the journal work as circular journal.

Reply:
Is the intention to use applied_to during replay to reduce the number of entries you need to apply?  How is this different from the tail of the journal?

Reply:
Yes, it is the same as the tail of the journal.